### PR TITLE
Rework `dlsym()`, `dlopen()`, `jl_dlsym()` and `jl_load_dynamic_library()` APIs

### PR DESCRIPTION
**Pull request by staticfloat**
* Remove `jl_dlsym_e()` to instead be rolled into `jl_dlsym()` with a `throw_err` parameter, similar to `jl_load_dynamic_library_()`.

* Fix #28881 by having `Libdl.dlsym()` return `nothing` on missing symbol, rather than `C_NULL`.

* Change `dlopen()` to mimic `dlsym()` so that a negative result returns `nothing`.  While not necessary in the same way as in the `dlsym()` case (there are no valid `NULL` `dlopen()` results) the consistency is worth it.

### CODE REVIEW

File: src/dlload.c [hunk 1]
```diff
@@ -192,56 +192,58 @@ static void *jl_load_dynamic_library_(const char *modname, unsigned flags, int t
...
-JL_DLLEXPORT void *jl_dlsym(void *handle, const char *symbol)
-{
-    void *ptr = jl_dlsym_e(handle, symbol);
-    if (!ptr)
-        jl_dlerror("could not load symbol \"%s\":\n%s", symbol);
-    return ptr;
+    /* Next, check for errors.  On Windows, a NULL pointer means the symbol
+     * was not found.  On everything else, we can have NULL symbols, so we check
+     * for non-NULL returns from dlerror().  */
+    const char * err = jl_dlerror();
```
[the remark this answers is not in the document]
Comment (staticfloat) [reply src/dlload.c]: Yes, I originally had it split out but it was enough extra code that I liked it better this way.  Optimizing for the programmer, not the program, I suppose.

File: src/dlload.c [hunk 2]
```diff
@@ -60,18 +60,18 @@ extern char *julia_bindir;
 
 #define JL_RTLD(flags, FLAG) (flags & JL_RTLD_ ## FLAG ? RTLD_ ## FLAG : 0)
 
-static void JL_NORETURN jl_dlerror(const char *fmt, const char *sym)
+static const char * jl_dlerror(void)
 {
 #ifdef _OS_WINDOWS_
-    CHAR reason[256];
+    static CHAR reason[256];
```
Comment (Keno) [inline src/dlload.c]: This makes this no longer thread safe, which is probably fine if we only do it in the error case, but if we do it in the success case as well, I'm worried the error message will get overwritten by the success of a different thread.

**Comment by Keno** [comment]
Do you want to separate out the `.c` change from the julia changes to make that part eligible for backporting and use from BinaryProvider?

File: src/ccall.cpp [hunk 3]
```diff
@@ -1490,13 +1491,11 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
     };
 #define is_libjulia_func(name) _is_libjulia_func((uintptr_t)&(name), #name)
 
+    static auto ptls_getter = &jl_get_ptls_states;
```
Comment (yuyichao) [inline src/ccall.cpp]: Why the change? This completely defeated the purpose of the old code, which is clearly stated in the comment..............

**Review by yuyichao — DISMISSED** [review]

File: src/ccall.cpp (hunk 3 above)
Comment (staticfloat) [reply src/ccall.cpp]: It's because I have been dealing with the dynamic (runtime) linker too much, so when it said "can cause linker issues" I naively thought that meant at runtime when the symbol actually gets used.  I didn't realize this caused issues at compile-time link.  :P

File: src/dlload.c (hunk 2 above)
Comment (Keno) [reply src/dlload.c]: Unfortunately, we can't assume that that's universally available.

Comment (yuyichao) [reply src/dlload.c]: If you really need to, you can use our own tls... There're already platform dependent members in there so it shouldn't be a huge problem even though it won't look as nice and will certainly needs cross references in the comments.

File: src/ccall.cpp (hunk 3 above)
Comment (yuyichao) [reply src/ccall.cpp]: You are welcome to change the comment in whichever way you want so that anyone that have read that comment won't change this code without testing the compiler configure combination again.....

File: src/dlload.c (hunk 2 above)
Comment (staticfloat) [reply src/dlload.c]: Because there seems to be some disagreement within the Julia codebase, I have taken the liberty of adding `#define JL_THREAD_LOCAL` to `julia.h` which emulates what is held within `src/support/dirname.c`.  If this shouldn't be done (and we should instead use the more complex TLS implementation in `src/threading.c`), it's probably best to transition both `jl_dlerror()` and `dirname()` to use that new stuff instead.

**Comment by staticfloat** [comment]
With regards to splitting out the `.c` changes and the `.jl` changes; I don't really see how they can be changed; the `.jl` changes are dependent upon the `.c` changes, and the `.c` changes are not backwards-compatible.

**Comment by staticfloat** [comment]
```
ERROR: could not load symbol "SymRefreshModuleList":
The operation completed successfully.
```

^^ Due to a missing semicolon.  Impressive.

**Comment by staticfloat** [comment]
Huzzah! We are passing all tests.  If I could get a thumbs up or thumbs down on the API exposed here, then we can make the final decision. 

I have purposefully not put in deprecation warnings for the *_e() methods in Julia, mostly because it would be very annoying for no real benefit at the moment. However it is something that should eventually be done, perhaps once we are closer to 2.0.

**Comment by Keno** [comment]
Ok by me. @yuyichao ?

File: src/ccall.cpp [hunk 4]
```diff
@@ -1490,13 +1491,15 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
     };
 #define is_libjulia_func(name) _is_libjulia_func((uintptr_t)&(name), #name)
 
+    static jl_ptls_t (*ptls_getter)(void);
+    // directly accessing the address of an ifunc can cause compile-time linker issues
+    // on some configurations (e.g. AArch64 + -Bsymbolic-functions), so we guard the
+    // `&jl_get_ptls_states` within this `#ifdef` guard, and use a more roundabout
+    // method involving `jl_dlsym()` on Linux platforms instead.
 #ifdef _OS_LINUX_
```
Comment (yuyichao) [inline src/ccall.cpp]: This actually makes the `static` useless. The performance isn't super critical here so if you want to just get rid of the `static` that's fine. Otherwise you can write it as.

```c++
static auto ptls_getter = [] {
#ifdef _OS_LINUX_
    jl_ptls_t (*p)(void);
    jl_dlsym(....);
    return p;
#else
    return jl_get_ptls_states;
#end
}();
```

File: src/dlload.c [hunk 5]
```diff
@@ -60,18 +60,18 @@ extern char *julia_bindir;
 
 #define JL_RTLD(flags, FLAG) (flags & JL_RTLD_ ## FLAG ? RTLD_ ## FLAG : 0)
 
-static void JL_NORETURN jl_dlerror(const char *fmt, const char *sym)
+static const char * jl_dlerror(void)
 {
 #ifdef _OS_WINDOWS_
-    CHAR reason[256];
+    static JL_THREAD_LOCAL CHAR reason[256];
```
Comment (yuyichao) [inline src/dlload.c]: No change needed but note that the "more complex" tls implementation we have should actually be simpler and faster.

**Review by yuyichao — COMMENTED** [review]
LGTM. The ccall ptls getter lookup is correct but is just very confusing though....

File: src/dlload.c (hunk 5 above)
Comment (yuyichao) [reply src/dlload.c]: Simpler in the code generated that is.

File: src/ccall.cpp (hunk 4 above)
Comment (staticfloat) [reply src/ccall.cpp]: Wow, that's available even all the way back to `GCC 4.8.5`.  I'm amazed.  And yes, you're right this completely defeats `static`, but I like this fancy lambda initialization.  I'm gonna go with that since compiler support is good enough.

Comment (yuyichao) [reply src/ccall.cpp]: This is standard C++11 feature and we require C++11 so that shouldn't be very surprising......

**Comment by staticfloat** [comment]
When this goes green, I'm going to merge.  Thanks for the shepherding, Keno and Yichao!

**Comment by vtjnash** [comment]
I'm not sure that I like the new API any better than the old, but I suppose that's not very helpful feedback.

**Comment by KristofferC** [comment]
Seems like the correct place to me.
